File: LeetCode/1872. Stone Game VIII/solution.cpp
// class Solution {
// public:
//     int stoneGameVIII(vector<int>& stones) {
//         vector<int> sums(stones.begin(), stones.end());
//         int n = stones.size();
//         for (int i = 1; i < n; i++)
//             sums[i] += sums[i-1];
//         vector<int> dp(n+1);
//         dp[n] = 0;
//         int g = INT_MIN;
//         for (int i = n-1; i >= 0; i--){
//             g = max(g, sums[i]-dp[i+1]);
//             dp[i] = g;
//         }
//         return dp[1];
//     }
// };

// class Solution {
// public:
//     int stoneGameVIII(vector<int>& stones) {
//         vector<int> sums(stones.begin(), stones.end());
//         int n = stones.size();
//         for (int i = 1; i < n; i++)
//             sums[i] += sums[i-1];
//         int dp = 0;
//         int g = INT_MIN;
//         for (int i = n-1; i > 0; i--){
//             g = max(g, sums[i]-dp);
//             dp = g;
//         }
//         return dp;
//     }
// };

class Solution {
public:
    int stoneGameVIII(vector<int>& stones) {
        int sums = 0, n = stones.size();
        for (int i = 0; i < n; i++)
            sums += stones[i];
        int dp = 0, g = INT_MIN;
        for (int i = n-1; i > 0; i--){
            g = max(g, sums-dp);
            dp = g;
            sums -= stones[i];
        }
        return dp;
    }
};

File: LeetCode/1872. Stone Game VIII/solution.py
class Solution:
    def stoneGameVIII(self, stones: List[int]) -> int:
        n = len(stones)
        for i in range(1,n):
            stones[i] += stones[i-1]
        f = 0
        g = stones[-1]-f
        for i in range(1,n)[::-1]:
            f = g 
            g = max(g, stones[i-1]-f)
        return f 

class Solution:
    def stoneGameVIII(self, stones: List[int]) -> int:
        n = len(stones)
        presums = [0]
        for stone in stones:
            presums.append(presums[-1]+stone)
        # if n == 2:
        #     return presums[-1]
        dp = [0]*n
        f = [-float('inf')]*n
        f[-1] = presums[n]-dp[n-1]
        for i in range(n-1)[::-1]:
            dp[i] = f[i+1]
            f[i] = max(f[i+1], presums[i+1]-dp[i])
        # print(presums)
        # print(dp)
        # print(f)
        return dp[0]
